### ISSUE TITLE
@BeforeInsert not firing on getRepository(<entity>).save()

### ISSUE THREAD

**Issue by nealdyrkacz** [typeorm/typeorm#5493]
**Issue type:**

[ X] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[X ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[X ] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

I have looked through the other issues with @BeforeInsert and it not firing on the save.  I have followed the instructions of the other answers but I have had no luck.

Creating a new instance of the Identity Entity
```
let identity: Identity = new Identity();
    identity.username = 'test';
    identity.password = 'password';
    await getRepository(Identity).save(identity);
```

Identity
```
@Entity()
export default class Identity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  username: string;

  @Column()
  password: string;

  @BeforeInsert()
  async generatePasswordHash(): Promise<void> {
    console.log('GENERATE');
    this.password = await bcrypt.hashSync(this.password, bcrypt.genSaltSync(10));
  }
```

What am I doing wrong?  I have tested both ways with making ```generatePasswordHash``` async and not. No luck.

**Comment by nealdyrkacz** [comment]
I was able to get subscribers to work though.

**Comment by TweededBadger** [comment]
I'm fairly sure when you call:
`await getRepository(Identity).save(identity);`
 `identity` is just getting used an object, so you are coming up against the issue here: https://github.com/typeorm/typeorm/issues/2924
This means that the instance of `Identity` that is created isn't actually instantiated, and therefore the `BeforeInsert` won't call. 

I think you'll need to run `await entityManager.save(identity);` instead.

**Comment by nealdyrkacz** [comment]
Hmmmm
`const entityManager = getManager();
const obj - entityManager.create(Obj)
await entityManger.save(obj)`
hits the @BeforeInsert and works.

while
`const entityManager = getManager();
const obj - new Obj();
await entityManger.save(obj)`

does not

**Comment by nealdyrkacz** [comment]
@TweededBadger , not sure of the difference.... the documentation says getManager().create() is the same as using new.  But thanks for pointing me in the right direction.

**Comment by ivawzh** [comment]
Same when using activerecord base entity

```ts
  ...
  @BeforeInsert()
  protected setCreatedAt() {
    this.createdAt = new Date()
  }
}

const meow = new Cat()
meow.save()
// Error: null value in column "created_at" violates not-null constraint
```

**Comment by iliyaZelenko** [comment]
I had a similar problem. I replaced this code:

``` js
this.ClipsRepository.save({
      title,
      start,
      end,
      author,
      description,
      videoId
    })
```

With:

``` js
this.ClipsRepository.save(
      this.ClipsRepository.create({
        title,
        start,
        end,
        author,
        description,
        videoId
      })
    )
```

And now the `@BeforeInsert()` is called.

**Comment by kuangshp** [comment]
if i code
await queryRunner.manager.insert<SponsorEntity>(SponsorEntity, {})
how will I trigger it @BeforeInsert()

**Comment by onichandame** [comment]
Any update on this?
If it's a bug, how and when are we expecting to fix it? If it's a feature, perhaps we should patch the documentation

**Comment by faridanthony** [comment]
Agreed. Ran into this on my project as well. A bit annoying having to call .create({...}) before every .save() or .update() to fire beforeInsert

**Comment by xdc0** [comment]
Looks like a limitation on the `EventListenerMetadata` class when checking if the object is eligible to have callbacks applied to it: https://github.com/typeorm/typeorm/blob/1197f881382d5d7a1f604ab53ae34e9461bf1ffe/src/metadata/EntityListenerMetadata.ts#L61

Plain objects will not pass the validation with the current implementation.

**Comment by SM2DevLLC** [comment]
Yeah any update on this? It's a mess that every bit of documentation on this is wrong. I've even been going through 3 different courses on this and all of them say different ways to update and none of them work lol.

**Comment by JesperBalslev** [comment]
Seriously, this feels beyond broken, what is the correct way to handle this issue?

**Comment by JoePotentier** [comment]
Any update on this? Just ran into it not encrypting passwords...

**Comment by hansanwok** [comment]
My case, just pass new Instance of User instead of createUserDto object to insert method
```
async create(createUserDto: CreateUserDto): Promise<InsertResult> {
    return this.usersRepository.insert(this.usersRepository.create(createUserDto))
  }
```

**Comment by ThatAnonyG** [comment]
> I was able to get subscribers to work though.

Hey can you tell me how you got that to work? I cannot get Subscribers to work with a similar setup where I am saving objects directly instead of instantiating them.

**Comment by makhloughi** [comment]
still broken? :|

**Comment by briermay** [comment]
sad part is, imo, this is as simple as going into their save call, and have IT automatically call create.

**Comment by mirupal** [comment]
For me adding `create`s for the hook BeforeInsert to be fired introduced one other problem.

The process stuck in an _infinite loop_, as I was setting an already loaded entity for a relation on the object to be created like this.

dtoObj.
It was a OneToMany relation, where I was trying to save an entity from the `Many` side.

Check out this Medium Post for in-depth analysis and solutions for an issue with `create` method: https://itnext.io/two-lines-change-turned-a-6-sec-request-to-300ms-cf0f13c00a75

So beware what you pass into the `create` method.

Solution from the Article:

```ts
const transaction = queryRunner.manager.create(Transaction, {
  seller: { id: sellerInstance.id },
  buyer: { id: buyerInstance.id },
});
```
